### ISSUE TITLE
Paypal Payments Pro IPN keeping payments marked as Pending Payment

### ISSUE THREAD

**Issue by quantumAV** [magento/magento2#25659]
### Preconditions (*)

1. Magento 2.3.3 
2.

### Steps to reproduce (*)

1. [https://github.com/magento/magento2/issues/18148]
2. On upgrading from 2.3.2 to 2.3.3 
@numan201 fix worked for me in 2.3.2 however on upgrading to 2.3.3 we now have Pending Payment status for orders received even though the details of the payment are on the order.
the fix that worked in 2.3.2 is already present in 2.3.3 - 

> For now, we have changed line 38 of vendor/magento/module-sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php from:
$state = $order->getState() ?: Order::STATE_PROCESSING; to $state = Order::STATE_PROCESSING; as it was before the update.

### Expected result (*)

1. payments from Paypal Express and Web Payments Pro should have a status of Processing. 

### Actual result (*)

1. Payments from Paypal Express are coming through correctly and have a status of Processing. Paypal Web Payments Pro are coming through as Pending Payment which is incorrect however the payment does show in the order information.

**Comment by magento-engcom-team** [comment]
@engcom-Echo Thank you for verifying the issue.

Unfortunately, not enough information was provided to acknowledge ticket. Please consider adding the following:
- [ ] Add `"Component: "` label(s) to this ticket based on verification result. If uncertain, you may follow the best guess
- [ ] Add `"Reproduced on "` label(s) to this ticket based on verification result

Once all required information is added, please add label `"Issue: Confirmed"` again.
Thanks!

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @engcom-Echo 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-29013` were created

**Issue Available**: @engcom-Echo, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by SlimJim192** [comment]
>     2. On upgrading from 2.3.2 to 2.3.3
>        @numan201 fix worked for me in 2.3.2 however on upgrading to 2.3.3 we now have Pending Payment status for orders received even though the details of the payment are on the order.
>        the fix that worked in 2.3.2 is already present in 2.3.3 -
> 
> 
> > For now, we have changed line 38 of vendor/magento/module-sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php from:
> > $state = $order->getState() ?: Order::STATE_PROCESSING; to $state = Order::STATE_PROCESSING; as it was before the update.

For myself the above fix was not present in 2.3.3 and the fix worked without problems. Perhaps you should double check?

**Comment by azambon** [comment]
Made a PR for this. Let's see if this finally gets something moving after over one year of nothingness.

**Comment by VladimirZaets** [comment]
Hi @quantumAV. Thank you for your report. 
The issue has been fixed in magento/magento2#25876 by @azambon in 2.4-develop branch
Related commit(s):
 - [45592e234b8ee700b02e82c28c9c7ee94ae13b51](https://github.com/magento/magento2/commit/45592e234b8ee700b02e82c28c9c7ee94ae13b51)

The fix will be available with the upcoming 2.4.0 release.

**Comment by azambon** [comment]
When I made the PR to fix this issue I targeted Magento 2.3 since at the time 2.4 had not been announced yet.
When 2.4 was announced I was confident that the PR would be backported to 2.3 as well. However version 2.3.4 has been released and the fix is not in there (see: https://github.com/magento/magento2/blob/2.3.4/app/code/Magento/Sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php ).

Is there any hope that the fix for this issue will be backported to 2.3, perhaps in 2.3.5?

**Comment by azambon** [comment]
2.3.5 is out and it still contains the bug fixed by this PR

https://github.com/magento/magento2/blob/2.3.5-p1/app/code/Magento/Sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php#L38

**Comment by tushhan** [comment]
I can confirm that the problem persists on Magento ver. 2.3.4 as well. Many orders display as "pending payment" and then "cancelled", customers call and complain while the money is actually transferred to paypal account. Payment/Order issues must be priority for Magento team while having the the same problem with recent versions.

**Comment by zigojacko** [comment]
So I am seeing that this issue is still apparent in version 2.3.5-p1 even after [countless reports stemming back a few years](https://github.com/magento/magento2/issues/18148) and reading the above, am I right in reading that this issue is also still apparent in version 2.4.0 ?

**Comment by azambon** [comment]
> So I am seeing that this issue is still apparent in version 2.3.5-p1 even after [countless reports stemming back a few years](https://github.com/magento/magento2/issues/18148) and reading the above, am I right in reading that this issue is also still apparent in version 2.4.0 ?

I haven't tried installing 2.4.0 to make sure, but the PR that I made got merged in 2.4 (not to be confused with 2.3.4). So I'd say that 2.4.0 should not be affected, but I stand to be corrected.
2.3.x is another matter, though. Despite this bug having been originally opened for 2.3.3 and despite the issue having been introduced back with 2.2.6, the PR has not been merged into any 2.3.x version that I know of.

So, to sum it up, 2.3.x should still contain the bug, while it should be fixed in 2.4.x

**Comment by zigojacko** [comment]
> > So I am seeing that this issue is still apparent in version 2.3.5-p1 even after [countless reports stemming back a few years](https://github.com/magento/magento2/issues/18148) and reading the above, am I right in reading that this issue is also still apparent in version 2.4.0 ?
> 
> I haven't tried installing 2.4.0 to make sure, but the PR that I made got merged in 2.4 (not to be confused with 2.3.4). So I'd say that 2.4.0 should not be affected, but I stand to be corrected.
> 2.3.x is another matter, though. Despite this bug having been originally opened for 2.3.3 and despite the issue having been introduced back with 2.2.6, the PR has not been merged into any 2.3.x version that I know of.
> 
> So, to sum it up, 2.3.x should still contain the bug, while it should be fixed in 2.4.x

Thanks for the clarification and feedback.

We've updated your commit for 2.4 in our 2.3.5 version which seems to work fine for the moment.

**Comment by kevinruscoe** [comment]
This happens in 2.2.11, but the code fix from https://github.com/magento/magento2/blob/2.3.5-p1/app/code/Magento/Sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php#L38 works fine. So I guess this fix can be pushed to the 2.2 branch?

**Comment by azambon** [comment]
> This happens in 2.2.11, but the code fix from https://github.com/magento/magento2/blob/2.3.5-p1/app/code/Magento/Sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php#L38 works fine. So I guess this fix can be pushed to the 2.2 branch?

I see no fix in 2.3.5-p1, only in 2.4.0.

If you do a "blame" of that file for version 2.3.5-p1 you'll find the commit that introduced the bug: https://github.com/magento/magento2/commit/f3234e1f59a282e26034a03f5d85cb24ba54bcd8#diff-68f9035e0545c5442dde19b0ba0fa9b4L38

Compare with 2.4.0: https://github.com/magento/magento2/blame/2.4.0/app/code/Magento/Sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php#L45

**Comment by azambon** [comment]
2.3.6 is out, and the fix is still missing from it

https://github.com/magento/magento2/blob/2.3.6/app/code/Magento/Sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php#L38

**Comment by zigojacko** [comment]
> 2.3.6 is out, and the fix is still missing from it
> 
> https://github.com/magento/magento2/blob/2.3.6/app/code/Magento/Sales/Model/Order/Payment/State/RegisterCaptureNotificationCommand.php#L38

Surprise surprise. 😂

**Comment by ihor-sviziev** [comment]
Hi,

This issue was already fixed in https://github.com/magento/magento2/pull/25876 and available in  Magento 2.4.0.
For everyone who wants to have this working on 2.3.x release line I would recommend to apply this fix as a patch, described in https://devdocs.magento.com/guides/v2.4/comp-mgr/patching.html.

I'm closing this issue as it's already solved.
